File: Advent2022/Day6/Day6.py
from typing import Optional

import aoc_utils as au


def distinct_start(input: str, number: int) -> Optional[int]:
    for i in range(number, len(input)):
        if len(set(input[i - number : i])) == number:
            return i

    return None


def test_one(input: str) -> Optional[int]:
    return distinct_start(input, 4)


# expected 1155
def part_one(input: str) -> Optional[int]:
    return distinct_start(input, 4)


def test_two(input: str) -> Optional[int]:
    return distinct_start(input, 14)


# expected 2789
def part_two(input: str) -> Optional[int]:
    return distinct_start(input, 14)


if __name__ == "__main__":
    day = au.Day(
        2022,
        6,
        test_one,
        test_two,
        part_one,
        part_two,
        test_input=[
            "mjqjpqmgbljsphdztnvjfqwrcgsmlb",
            "bvwbjplbgvbhsrlpgdmjqwftvncz",
            "nppdvjthqldpwncqszvftbrmjlhg",
            "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
            "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw",
        ],
        input=au.OneLineParser(),
    )

    day.run_all(run_tests=True)

File: Advent2022/Day6/Day6.cpp
#include "aoc_utils.h"

#include <iostream>
#include <string>
#include <set>

using namespace std;

class Day6 : public au::Day<vector<string>>
{
private:
public:
    Day6() : Day(2022, 6) {}

    int distinct_start(string input, unsigned int number)
    {
        for (unsigned int i = number; i < input.length(); i++)
            if (set<char>(input.begin() + (i - number), input.begin() + i).size() == number)
                return i;

        return -1;
    }

    int test_one(const vector<string> &input)
    {
        for (string s : input)
            cout << "Test: " << distinct_start(s, 4) << "   <- [ " << s << " ]" << endl;
        return 0;
    }

    int part_one(const vector<string> &input)
    {
        return distinct_start(input[0], 4);
    }

    int test_two(const vector<string> &input)
    {
        for (string s : input)
            cout << "Test: " << distinct_start(s, 14) << "   <- [ " << s << " ]" << endl;
        return 0;
    }

    int part_two(const vector<string> &input)
    {
        return distinct_start(input[0], 14);
    }
};

int main()
{
    Day6().run_all(true);

    return 0;
}
